### PR TITLE
Modify Transpose kernel to work in TFLu

### PR DESCRIPTION
**Pull request by patriklaurell**
This PR modifies the transpose kernel to make it run in micro, ports the tests and adds the kernel to the micro build.

This is PR 4/4 in delivering #45695

### CODE REVIEW

**Comment by abattery** [comment]
This PR also impacts the performance changes on the TFLite. Could you use the micro macro to divide the logics? For example, keeping the original logic for the TFlite and using a new logics for only micro cases.

**Comment by patriklaurell** [comment]
@abattery I am not sure I understand what you mean. The changes in this PR should only affect the micro kernel. How does this impact the performance changes on TFLite?

**Comment by abattery** [comment]
Sorry @patriklaurell you are right. I misread the filename. Please ignore my comment.

**Comment by dmpiergiacomo** [comment]
Hi, which is the status of this PR? Thank you.

**Comment by patriklaurell** [comment]
@petewarden I updated this PR and added two commits for porting the tests and adding transpose to the micro build. It is ready for review.

**Comment by patriklaurell** [comment]
@advaitjain @petewarden The MacOS CPU Python3 test is failing but there is very little information in the test log and I have no way of reproducing the error locally. Do you have any insight into why it fails?

File: tensorflow/lite/micro/kernels/transpose.cc [hunk 1]
```diff
@@ -80,71 +45,60 @@ TfLiteStatus Prepare(TfLiteContext* context, TfLiteNode* node) {
...
-#define TF_LITE_TRANSPOSE(type, scalar)                     \
-  type::Transpose(params, GetTensorShape(op_context.input), \
-                  GetTensorData<scalar>(op_context.input),  \
-                  GetTensorShape(op_context.output),        \
-                  GetTensorData<scalar>(op_context.output))
+#define TF_LITE_TRANSPOSE(type)                                      \
+  reference_ops::Transpose(params, GetTensorShape(op_context.input), \
+                           GetTensorData<type>(op_context.input),    \
+                           GetTensorShape(op_context.output),        \
+                           GetTensorData<type>(op_context.output))
```
Comment (advaitjain) [inline tensorflow/lite/micro/kernels/transpose.cc:77]: remove this #define and directly call the reference function from the switch case.

File: tensorflow/lite/micro/kernels/transpose.cc [hunk 2]
```diff
@@ -80,71 +45,60 @@ TfLiteStatus Prepare(TfLiteContext* context, TfLiteNode* node) {
...
 
-  // Transpose kernel only does rearranging values not numeric evaluations on
-  // each cell. It's safe to implement per size of scalar type and this trick
-  // keeps the total code size in a reasonable range.
+  // Transpose kernel only does rearranging values not numeric evaluations
+  // on each cell. It's safe to implement per size of scalar type and this
+  // trick keeps the total code size in a reasonable range.
   switch (op_context.input->type) {
```
Comment (advaitjain) [inline tensorflow/lite/micro/kernels/transpose.cc:82]: only add support for int8 and float32 for the initial port PR. Everything else should be done separately on an as-needed basis.

**Review by advaitjain — CHANGES_REQUESTED** [review]

**Comment by advaitjain** [comment]
> @advaitjain @petewarden The MacOS CPU Python3 test is failing but there is very little information in the test log and I have no way of reproducing the error locally. Do you have any insight into why it fails?

Unfortunately, I don't know why it is failing either. Let's have you address the review comments and then we can try and figure out what is going on with the MacOS build.

**Comment by patriklaurell** [comment]
@advaitjain ready for review

**Review by advaitjain — APPROVED** [review]

**Comment by advaitjain** [comment]
> @advaitjain ready for review. Did you find anything regarding the MacOS test?

I really do not know what is going on with the MacOS failure. @petewarden, any ideas?

**Comment by advaitjain** [comment]
Can you merge master and then get this PR building again after changes from https://github.com/tensorflow/tensorflow/commit/822ddbf71817de64df9b6741e007a240e7b61c8d

That will fix the MacOS and internal CI errors.

**Pull request merged**